platform/pulp/platform/tasks/task_system.py

```diff
@@ -1,4 +1,548 @@
+    for task_status in Worker.objects.tasks.filter(state__in=constants.CALL_INCOMPLETE_STATES):
+        cancel(task_status.id)
+
+    # Delete the worker document
+    Worker.objects.get(name=name).delete()
+
+    # If the worker is a resource manager, we also need to delete the associated lock
+    if name.startswith(RESOURCE_MANAGER_WORKER_NAME):
+        ResourceManagerLock.objects.filter(name=name).delete()
+
+
+@task(base=PulpTask)
+def _release_resource(task_id):
+    """
+    Do not queue this task yourself. It will be used automatically when your task is dispatched by
+    the _queue_reserved_task task.
+
+    When a resource-reserving task is complete, this method releases the resource by removing the
+    ReservedResource object by UUID.
+
+    :param task_id: The UUID of the task that requested the reservation
+    :type  task_id: basestring
+    """
+    try:
+        TaskStatus.objects.get(id=task_id, state=constants.CALL_RUNNING_STATE)
+    except TaskStatus.DoesNotExist:
+        pass
+    else:
+        new_task = Task()
+        msg = _('The task status %(task_id)s exited immediately for some reason. Marking as '
+                'errored. Check the logs for more details')
+        runtime_exception = RuntimeError(msg % {'task_id': task_id})
+
+        class MyEinfo(object):
+            traceback = None
+
+        new_task.on_failure(runtime_exception, task_id, (), {}, MyEinfo)
+
+    ReservedResource.objects.filter(task__id=task_id).delete()
+
+
+class TaskResult(object):
+    """
+    The TaskResult object is used for returning errors and spawned tasks that do not affect the
+    primary status of the task.
+
+    Errors that don't affect the current task status might be related to secondary actions
+    where the primary action of the async-task was successful
+
+    Spawned tasks are items such as the individual tasks for updating the bindings on
+    each consumer when a repo distributor is updated.
+    """
+
+    def __init__(self, result=None, error=None, spawned_tasks=None):
+        """
+        :param result: The return value from the task
+        :type result: dict
+        :param error: The PulpException for the error & sub-errors that occurred
+        :type error: pulp.server.exception.PulpException
+        :param spawned_tasks: A list of task status objects for tasks that were created by this
+                              task and are tracked through the pulp database.
+                              Alternately an AsyncResult, or the task_id of the task created.
+        :type spawned_tasks: list of TaskStatus, AsyncResult, or str objects
+        """
+        self.return_value = result
+        self.error = error
+        self.spawned_tasks = []
+        if spawned_tasks:
+            for spawned_task in spawned_tasks:
+                if isinstance(spawned_task, dict):
+                    self.spawned_tasks.append({'task_id': spawned_task.get('task_id')})
```

Comment:
i do not really know where this dict comes from but make sure that it has 'task_id' key and not just 'id'